File: 547-number-of-provinces/number-of-provinces.cpp
class Solution {
public:
    void dfs(int i,vector<vector<int>>&adj,vector<int>&vis){
        vis[i]=1;
        for(auto x:adj[i]){
            if(!vis[x]){
                dfs(x,adj,vis);
            }
        }
    }
    int findCircleNum(vector<vector<int>>& m) {
        int n=m.size();
        vector<vector<int>>adj(n);
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(m[i][j]==1 && i!=j){
                    adj[i].push_back(j);
                    adj[j].push_back(i);
                }
            }
        }
       int cnt=0;
       vector<int>vis(n,0);
        for(int i=0;i<n;i++){
            if(!vis[i]){
                dfs(i,adj,vis);
                cnt++;
            }
        }
        return cnt;
    }
};

File: 547-number-of-provinces/number-of-provinces.java
class Solution {
    public void bfs(ArrayList<ArrayList<Integer>> adj,int s,int[] vis) {
        Queue<Integer> q = new LinkedList<>();
        int n=adj.size();
        q.add(s);
        vis[s]=1;
        while(!q.isEmpty()){
            int node = q.poll();
            
            for(int nbr : adj.get(node)){
                if(vis[nbr]==0){
                    vis[nbr]=1;
                    q.add(nbr);
                }
            }
            
        }
    }
    public int findCircleNum(int[][] isConnected) {
        int n=isConnected.length;
        ArrayList<ArrayList<Integer>>adj = new ArrayList<>();
        for(int i=0;i<n;i++){
            adj.add(new ArrayList<>());
        }
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(i!=j && isConnected[i][j]==1){
                    adj.get(i).add(j);
                    adj.get(j).add(i);
                }
            }
        }
        int cnt=0;
        int[] vis = new int[n];
        for(int i=0;i<n;i++){
            if(vis[i]==0){
                bfs(adj,i,vis);
                cnt++;
            }
        }
        return cnt;


    }
}